gtars-overlaprs/src/genome_index.rs

```diff
@@ -0,0 +1,678 @@
+//!
+//! # Examples
+//!
+//! ## Basic Usage
+//!
+//! ```
+//! use gtars_overlaprs::{OverlapperType, genome_index::IntoGenomeIndex};
+//! use gtars_core::models::{Region, RegionSet};
+//!
+//! // Create a genome-wide set of intervals (e.g., gene annotations)
+//! let genes = vec![
+//!     Region { chr: "chr1".to_string(), start: 1000, end: 2000, rest: Some("BRCA1".to_string()) },
+//!     Region { chr: "chr1".to_string(), start: 5000, end: 6000, rest: Some("TP53".to_string()) },
+//!     Region { chr: "chr2".to_string(), start: 1000, end: 3000, rest: Some("EGFR".to_string()) },
+//! ];
+//!
+//! let gene_set = RegionSet::from(genes);
+//! let genome_index = gene_set.into_genome_index(OverlapperType::AiList);
+//!
+//! // Query multiple regions across different chromosomes
+//! let query_regions = RegionSet::from(vec![
+//!     Region { chr: "chr1".to_string(), start: 1500, end: 2500, rest: None },
+//!     Region { chr: "chr2".to_string(), start: 2000, end: 4000, rest: None },
+//! ]);
+//!
+//! let overlaps = genome_index.find_overlaps(&query_regions);
+//! println!("Found {} overlapping features", overlaps.len());
+//! ```
+//!
+//! ## Iterator-Based Query
+//!
+//! For memory-efficient processing of large result sets:
+//!
+//! ```
+//! use gtars_overlaprs::{OverlapperType, genome_index::IntoGenomeIndex};
+//! use gtars_core::models::{Region, RegionSet};
+//!
+//! # let genes = vec![
+//! #     Region { chr: "chr1".to_string(), start: 1000, end: 2000, rest: Some("BRCA1".to_string()) },
+//! # ];
+//! # let gene_set = RegionSet::from(genes);
+//! # let genome_index = gene_set.into_genome_index(OverlapperType::AiList);
+//! # let query_regions = RegionSet::from(vec![
+//! #     Region { chr: "chr1".to_string(), start: 1500, end: 2500, rest: None },
+//! # ]);
+//! // Process overlaps one at a time without allocating a vector
+//! for (chr, interval) in genome_index.find_overlaps_iter(&query_regions) {
+//!     println!("Chromosome {}: {:?}", chr, interval);
+//! }
+//! ```
+
+use std::{collections::HashMap, fmt::Debug};
+
+use gtars_core::models::{Interval, Region, RegionSet};
+use num_traits::{PrimInt, Unsigned};
+use thiserror::Error;
+
+use crate::{AiList, Bits, Overlapper, OverlapperType};
+
+/// Errors that can occur when working with [`GenomeIndex`].
+#[derive(Debug, Error)]
+pub enum GenomeIndexError {
+    /// Error parsing a genomic region string.
+    #[error("Error parsing region: {0}")]
+    RegionParsingError(String),
+    /// Error converting interval coordinates to the required integer type.
+    #[error("Error converting interval coordinates to u32: start={0}, end={1}")]
+    CoordinateConversionError(String, String),
+}
+
+/// A genome-wide index for efficient overlap queries across multiple chromosomes.
+///
+/// `GenomeIndex` maintains a separate overlap data structure (either [`AiList`] or [`Bits`])
+/// for each chromosome, enabling efficient queries across the entire genome. This is particularly
```

Comment:
The last sentence here doesn't really make sense: 

finding overlaps between genomic features on different chromosomes is impossible by definition.